python/ray/dag/compiled_dag_node.py

```diff
@@ -943,11 +929,61 @@ def _preprocess(self) -> None:
                 task.arg_type_hints.append(upstream_task.dag_node.type_hint)
 
                 if upstream_task.dag_node.type_hint.requires_nccl():
-                    # Add all readers to the NCCL actors of P2P.
-                    nccl_actors_p2p.add(downstream_actor_handle)
+                    custom_nccl_group: Optional[
+                        GPUCommunicator
+                    ] = upstream_task.dag_node.type_hint.get_custom_nccl_group()
+                    if custom_nccl_group is None:
+                        custom_nccl_group = self._default_nccl_group_p2p
+                    custom_nccl_group_to_p2p_senders_and_receivers[
+                        custom_nccl_group
+                    ].add(dag_node)
+
+        self._init_nccl_groups(
+            custom_nccl_group_to_p2p_senders,
+            custom_nccl_group_to_p2p_senders_and_receivers,
+            nccl_collective_ops,
+        )
 
-        nccl_actors_p2p = list(nccl_actors_p2p)
-        if None in nccl_actors_p2p:
+        if direct_input:
+            self._input_num_positional_args = 1
+        elif not input_positional_args:
+            self._input_num_positional_args = 0
+        else:
+            self._input_num_positional_args = max(input_positional_args) + 1
+        self._input_kwargs = tuple(input_kwargs)
+
+    def _init_nccl_groups(
+        self,
+        custom_nccl_group_to_p2p_senders: Dict[
+            Optional[GPUCommunicator], Set["ray.dag.DAGNode"]
+        ],
+        custom_nccl_group_to_p2p_senders_and_receivers: Dict[
+            Optional[GPUCommunicator], Set["ray.dag.DAGNode"]
+        ],
+        nccl_collective_ops: Set["ray.dag.collective_node._CollectiveOperation"],
+    ):
+        """
+        Initialize all NCCL groups used in the DAG.
+        1. For each custom NCCL group, verify that it contains the correct set of
+           actors and initialize it.
+        2. For each NCCL operation for which no NCCL group was specified, try to
+           reuse a NCCL group initialized in step 1.
+        3. If no NCCL group could be reused in step 2, construct a `_NcclGroup`.
```

Comment:
Seems that logic for collectives' nccl groups is not described?

Reply:
This is a generic description for both P2P and collectives. I changed the comment to clarify that NCCL operations include both P2P and collectives.